Add support for runtime field types to mappings editor.

cjcenizal opened the pull request:
Fixes https://github.com/elastic/kibana/issues/77110

To test, add a runtime field to an index template's mappings via the UI or via Console like so:

```
PUT /my-index/_mappings
{
    "properties" : {
        "day_of_week" : {
            "type" : "runtime",
            "runtime_type" : "keyword",
            "script" : {
                "source" : "emitValue(doc['timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
            }
        }
    }
}
```

## Release note

Index templates can now be configured with runtime fields in their mappings.

![image](https://user-images.githubusercontent.com/1238659/94231265-7e064880-feb8-11ea-9033-be60c9b527e4.png)

![image](https://user-images.githubusercontent.com/1238659/94231286-878fb080-feb8-11ea-947e-5c7d252c7de0.png)

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

cjcenizal commented on x-pack/plugins/index_management/public/application/components/mappings_editor/constants/data_types_definition.tsx:
```diff
@@ -13,6 +13,23 @@ import { documentationService } from '../../../services/documentation';
+      defaultMessage: 'Runtime',
+    }),
+    documentation: {
+      main: '/runtime_field.html',
```
Note that these docs don't exist yet.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/)
* Commit: 02e43dda58da1806ab3ee98e474ae47071acb449
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/execution/node/378/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/","number":"74328","commit":"02e43dda58da1806ab3ee98e474ae47071acb449"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/)
* Commit: 1e7b0fc51f5e6844f51c0876db4683ca2039114e

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/1e7b0fc51f5e6844f51c0876db4683ca2039114e) | diff | [baseline](https://github.com/elastic/kibana/commit/dcd119ce5f2935bcc5a027c45a51dfd29d1643a2) |
| --- | --- | --- | --- |
| indexManagement | 506 | +3 | 503 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/1e7b0fc51f5e6844f51c0876db4683ca2039114e) | diff | [baseline](https://github.com/elastic/kibana/commit/dcd119ce5f2935bcc5a027c45a51dfd29d1643a2) |
| --- | --- | --- | --- |
| indexManagement | 1.6MB | +5.3KB | 1.5MB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/1e7b0fc51f5e6844f51c0876db4683ca2039114e) | diff | [baseline](https://github.com/elastic/kibana/commit/dcd119ce5f2935bcc5a027c45a51dfd29d1643a2) |
| --- | --- | --- | --- |
| indexManagement | 263.3KB | +158.0B | 263.2KB |

### History
* :broken_heart: [Build #74328](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/) failed 02e43dda58da1806ab3ee98e474ae47071acb449

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/","number":"74328","commit":"02e43dda58da1806ab3ee98e474ae47071acb449"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/","number":"74447","commit":"1e7b0fc51f5e6844f51c0876db4683ca2039114e"}]}
PIPELINE-->

alisonelizabeth commented on x-pack/plugins/index_management/public/application/index.tsx:
```diff
@@ -44,17 +48,19 @@ export const renderApp = (
-          </MappingsEditorProvider>
-        </AppContextProvider>
-      </Provider>
+      <KibanaReactContextProvider>
```
I might be missing something here, but is this necessary? I don't see it being used anywhere.

cjcenizal replied:
The `CodeEditor` component used for editing the painless script requires it to provide `uiSettings`, which is really unclear from the code. I'll add a comment.

alisonelizabeth replied:
Ah. Yeah, it isn't very clear, but I see that now after looking at the code_editor code. Thanks for adding the comment!

alisonelizabeth reviewed (commented):
Thanks for working on this @cjcenizal!

I left a few comments in the code. 

I'm also wondering about the general workflow. For other field types that have required fields (such as `alias` and `token_count`), we require the required fields before adding it.

For example:
<img width="1272" alt="Screen Shot 2020-09-15 at 10 52 38 AM" src="https://user-images.githubusercontent.com/5226211/93228816-ec723a80-f743-11ea-8fec-43482a54228f.png">

I think we probably want to do the same for `runtime` field types. As it is, if I add a `runtime` field type via the UI first, I see this:

<img width="437" alt="Screen Shot 2020-09-15 at 10 51 25 AM" src="https://user-images.githubusercontent.com/5226211/93228981-19bee880-f744-11ea-8ea5-950034d9224c.png">

I'm also able to proceed to the review step, but end up getting an error back from ES since `runtime_type` and `script` are both required.

For the `runtime_type` parameter, I think we could possibly do something similar for `numeric` and `range` types, and have a subtype dropdown.

For example:

<img width="1205" alt="Screen Shot 2020-09-15 at 11 16 20 AM" src="https://user-images.githubusercontent.com/5226211/93229870-124c0f00-f745-11ea-80c5-54eee8ef3543.png">

Let me know what you think!

cjcenizal commented:
Thanks for the great review @alisonelizabeth! I incorporated your suggestions, and added a screenshot to the PR description to show how the UI now requires you to define the type before adding a runtime field. Could you please review again?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/)
* Commit: ef9e325f9365d824b9f8b4cc0284176560483970

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/ef9e325f9365d824b9f8b4cc0284176560483970) | diff | [baseline](https://github.com/elastic/kibana/commit/dcd119ce5f2935bcc5a027c45a51dfd29d1643a2) |
| --- | --- | --- | --- |
| indexManagement | 507 | +4 | 503 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/ef9e325f9365d824b9f8b4cc0284176560483970) | diff | [baseline](https://github.com/elastic/kibana/commit/dcd119ce5f2935bcc5a027c45a51dfd29d1643a2) |
| --- | --- | --- | --- |
| indexManagement | 1.6MB | +5.5KB | 1.5MB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/ef9e325f9365d824b9f8b4cc0284176560483970) | diff | [baseline](https://github.com/elastic/kibana/commit/dcd119ce5f2935bcc5a027c45a51dfd29d1643a2) |
| --- | --- | --- | --- |
| indexManagement | 263.3KB | +158.0B | 263.2KB |

### History
* :green_heart: [Build #74447](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/) succeeded 1e7b0fc51f5e6844f51c0876db4683ca2039114e
* :broken_heart: [Build #74328](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/) failed 02e43dda58da1806ab3ee98e474ae47071acb449

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/","number":"74328","commit":"02e43dda58da1806ab3ee98e474ae47071acb449"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/","number":"74447","commit":"1e7b0fc51f5e6844f51c0876db4683ca2039114e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/","number":"74653","commit":"ef9e325f9365d824b9f8b4cc0284176560483970"}]}
PIPELINE-->

alisonelizabeth reviewed (commented):
Thanks for making the changes @cjcenizal! This is looking great.

`script` also appears to be a required field. I understand the editor takes up some real estate, but I think it's important to make the user provide a value here before being allowed to submit the form. WDYT?

Also, I had a question about the copy for `keyword` and `ip` runtime types. 

<img width="718" alt="Screen Shot 2020-09-16 at 10 17 19 AM" src="https://user-images.githubusercontent.com/5226211/93349991-273ca700-f806-11ea-8986-008aac44a8fc.png">
<img width="713" alt="Screen Shot 2020-09-16 at 10 17 37 AM" src="https://user-images.githubusercontent.com/5226211/93349994-27d53d80-f806-11ea-84af-0b68eddc48c8.png">

Is "...use the text data type" and "...use the IP range data type" still relevant in the context of runtime fields?

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/)
* Commit: 9af93bc1ff710343a310a7601943f73973053077
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Build metrics
> ‼️ ERROR: build [4773d17f-6524-4650-a331-dc706e5d5433] does not have valid git info

### History
* :green_heart: [Build #74653](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/) succeeded ef9e325f9365d824b9f8b4cc0284176560483970
* :green_heart: [Build #74447](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/) succeeded 1e7b0fc51f5e6844f51c0876db4683ca2039114e
* :broken_heart: [Build #74328](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/) failed 02e43dda58da1806ab3ee98e474ae47071acb449

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/","number":"74328","commit":"02e43dda58da1806ab3ee98e474ae47071acb449"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/","number":"74447","commit":"1e7b0fc51f5e6844f51c0876db4683ca2039114e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/","number":"74653","commit":"ef9e325f9365d824b9f8b4cc0284176560483970"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/","number":"76036","commit":"9af93bc1ff710343a310a7601943f73973053077"}]}
PIPELINE-->

cjcenizal commented:
> `script` also appears to be a required field. I understand the editor takes up some real estate, but I think it's important to make the user provide a value here before being allowed to submit the form. WDYT?

Whoops, good catch! Thanks for calling this out. I added validation. I had to surface the error message as a callout in addition to an error text, since the code editor is tall enough that it pushes the error text out of view on shorter screens. I think we need to solve this in the code editor component, e.g. by making it resize vertically to accommodate its content.

![image](https://user-images.githubusercontent.com/1238659/93828625-978d6300-fc20-11ea-8916-93268e91db44.png)

![image](https://user-images.githubusercontent.com/1238659/93828630-99572680-fc20-11ea-9cb6-162f333dbb92.png)

I also update a couple other places where we were using callouts similarly, to follow the patterns shown in the [EUI examples](https://elastic.github.io/eui/#/display/callout).

![image](https://user-images.githubusercontent.com/1238659/93828734-d1f70000-fc20-11ea-9763-0f813b213ef9.png)

![image](https://user-images.githubusercontent.com/1238659/93828738-d4595a00-fc20-11ea-8bab-51365348953d.png)

> Is "...use the text data type" and "...use the IP range data type" still relevant in the context of runtime fields?

This is a great question. It's difficult for me to answer because I have a hard time understanding the user's intention when defining runtime fields in the index template. Ostensibly, if the user is defining or editing an index template, they have some level of control over the ingestion system, which includes the source of the documents being ingested and the ingest node pipeline used to process them. If this is the case, then they might be making decisions about how best to store, say, an IP. Maybe they're still trying to figure out the right design for their ingestion system, and maybe this text helps them realize they actually do want an IP range instead of an IP? Or maybe they started with the intention to define an IP range, but they didn't realize it's not yet supported by runtime fields? 🤷  I guess I'm not entirely sure if it's helpful, but I don't see how it hurts. Do you find it confusing?

cjcenizal commented:
@elasticmachine merge upstream

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/)
* Commit: 8a53f6850cbaa8f4948e938b4e1d1794c0a9fada
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-accessibility](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/execution/node/587/log)
* [Execute xpack-accessibility](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/execution/node/685/log)

### History
* :broken_heart: [Build #76036](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/) failed 9af93bc1ff710343a310a7601943f73973053077
* :green_heart: [Build #74653](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/) succeeded ef9e325f9365d824b9f8b4cc0284176560483970
* :green_heart: [Build #74447](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/) succeeded 1e7b0fc51f5e6844f51c0876db4683ca2039114e
* :broken_heart: [Build #74328](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/) failed 02e43dda58da1806ab3ee98e474ae47071acb449

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/","number":"74328","commit":"02e43dda58da1806ab3ee98e474ae47071acb449"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/","number":"74447","commit":"1e7b0fc51f5e6844f51c0876db4683ca2039114e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/","number":"74653","commit":"ef9e325f9365d824b9f8b4cc0284176560483970"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/","number":"76036","commit":"9af93bc1ff710343a310a7601943f73973053077"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/","number":"76039","commit":"8a53f6850cbaa8f4948e938b4e1d1794c0a9fada"}]}
PIPELINE-->

alisonelizabeth reviewed (approved):
Thanks for making the changes @cjcenizal. Great work! 

I still wonder if the `script` code editor should be present when initially adding the field (since it's required), but I'll leave that up to you.

(in this view:)
<img width="1190" alt="Screen Shot 2020-09-21 at 8 58 35 PM" src="https://user-images.githubusercontent.com/5226211/93835543-91fa4200-fc4d-11ea-8ad6-6967b5bafd52.png">

---

> This is a great question. It's difficult for me to answer because I have a hard time understanding the user's intention when defining runtime fields in the index template.

👍 I think it is fine to leave as-is. I was also wondering what the use case would look like.

cjcenizal commented:
@alisonelizabeth Good point. I was originally on the fence about adding the code editor to the "add field" component because I thought its height would disrupt the UX by causing content below it to jump around. But I think it's more important to be consistent with required parameters, so I added it in.  I made the editor shorter, which I think helps (and also allowed me to remove the callout I had put in). I doubt these scripts will be so involved that they require as much height as I originally provided.

![image](https://user-images.githubusercontent.com/1238659/93908363-60b45d00-fcb3-11ea-9a1a-9c3b10a35f72.png)

![image](https://user-images.githubusercontent.com/1238659/93908371-61e58a00-fcb3-11ea-9af8-62cb76a6dde7.png)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76232/)
* Commit: 00573c153579d9cbe65509e1ab297ee9df7a8f79
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76232/execution/node/377/log)

### History
* :broken_heart: [Build #76039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/) failed 8a53f6850cbaa8f4948e938b4e1d1794c0a9fada
* :broken_heart: [Build #76036](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/) failed 9af93bc1ff710343a310a7601943f73973053077
* :green_heart: [Build #74653](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/) succeeded ef9e325f9365d824b9f8b4cc0284176560483970
* :green_heart: [Build #74447](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/) succeeded 1e7b0fc51f5e6844f51c0876db4683ca2039114e
* :broken_heart: [Build #74328](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/) failed 02e43dda58da1806ab3ee98e474ae47071acb449

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74328/","number":"74328","commit":"02e43dda58da1806ab3ee98e474ae47071acb449"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/","number":"74447","commit":"1e7b0fc51f5e6844f51c0876db4683ca2039114e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/","number":"74653","commit":"ef9e325f9365d824b9f8b4cc0284176560483970"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/","number":"76036","commit":"9af93bc1ff710343a310a7601943f73973053077"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/","number":"76039","commit":"8a53f6850cbaa8f4948e938b4e1d1794c0a9fada"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76232/","number":"76232","commit":"00573c153579d9cbe65509e1ab297ee9df7a8f79"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76239/)
* Commit: 3ac63a6a861d3d01756203fa4722e041e41e3d5d
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-accessibility](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76239/execution/node/587/log)
* [Execute xpack-accessibility](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76239/execution/node/685/log)

### History
* :broken_heart: [Build #76232](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76232/) failed 00573c153579d9cbe65509e1ab297ee9df7a8f79
* :broken_heart: [Build #76039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/) failed 8a53f6850cbaa8f4948e938b4e1d1794c0a9fada
* :broken_heart: [Build #76036](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/) failed 9af93bc1ff710343a310a7601943f73973053077
* :green_heart: [Build #74653](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/) succeeded ef9e325f9365d824b9f8b4cc0284176560483970
* :green_heart: [Build #74447](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/) succeeded 1e7b0fc51f5e6844f51c0876db4683ca2039114e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74447/","number":"74447","commit":"1e7b0fc51f5e6844f51c0876db4683ca2039114e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/","number":"74653","commit":"ef9e325f9365d824b9f8b4cc0284176560483970"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/","number":"76036","commit":"9af93bc1ff710343a310a7601943f73973053077"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/","number":"76039","commit":"8a53f6850cbaa8f4948e938b4e1d1794c0a9fada"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76232/","number":"76232","commit":"00573c153579d9cbe65509e1ab297ee9df7a8f79"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76239/","number":"76239","commit":"3ac63a6a861d3d01756203fa4722e041e41e3d5d"}]}
PIPELINE-->

sebelga commented:
@cjcenizal Can you explain why you are not using 2 columns for the required fields? I am surprised to see a huge select occupying the whole width.  I understand in smaller screens those 2 will stack on top of another but on desktop?

Or maybe have a described group with the "Emited type" on the left and the select on the right. Then the same for the script.

cjcenizal commented:
Great idea, @sebelga! Done.

sebelga commented:
Awesome thanks @cjcenizal ! 

I haven't played with it yet but maybe in the future for those scenarios, it would be nice to make use of the [Resize observer](https://elastic.github.io/eui/#/utilities/resize-observer) and CSS (or maybe set the `stack` var in your case) to change the layout depending of the width of the parent.

This would allow us to create components with a layout that adapts to wherever they are thrown in.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77188/)
* Commit: 1b01e7ef9453503423051217c6dbb843d23b0751

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/1b01e7ef9453503423051217c6dbb843d23b0751) | diff | [baseline](https://github.com/elastic/kibana/commit/e53fcbff66508a020f26f0940e56de8b7a4379f5) |
| --- | --- | --- | --- |
| indexManagement | 508 | +5 | 503 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/1b01e7ef9453503423051217c6dbb843d23b0751) | diff | [baseline](https://github.com/elastic/kibana/commit/e53fcbff66508a020f26f0940e56de8b7a4379f5) |
| --- | --- | --- | --- |
| indexManagement | 1.6MB | +8.6KB | 1.6MB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/1b01e7ef9453503423051217c6dbb843d23b0751) | diff | [baseline](https://github.com/elastic/kibana/commit/e53fcbff66508a020f26f0940e56de8b7a4379f5) |
| --- | --- | --- | --- |
| indexManagement | 122.8KB | +158.0B | 122.6KB |

### History
* :broken_heart: [Build #76239](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76239/) failed 3ac63a6a861d3d01756203fa4722e041e41e3d5d
* :broken_heart: [Build #76232](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76232/) failed 00573c153579d9cbe65509e1ab297ee9df7a8f79
* :broken_heart: [Build #76039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/) failed 8a53f6850cbaa8f4948e938b4e1d1794c0a9fada
* :broken_heart: [Build #76036](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/) failed 9af93bc1ff710343a310a7601943f73973053077
* :green_heart: [Build #74653](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/) succeeded ef9e325f9365d824b9f8b4cc0284176560483970

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74653/","number":"74653","commit":"ef9e325f9365d824b9f8b4cc0284176560483970"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76036/","number":"76036","commit":"9af93bc1ff710343a310a7601943f73973053077"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76039/","number":"76039","commit":"8a53f6850cbaa8f4948e938b4e1d1794c0a9fada"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76232/","number":"76232","commit":"00573c153579d9cbe65509e1ab297ee9df7a8f79"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76239/","number":"76239","commit":"3ac63a6a861d3d01756203fa4722e041e41e3d5d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77188/","number":"77188","commit":"1b01e7ef9453503423051217c6dbb843d23b0751"}]}
PIPELINE-->
